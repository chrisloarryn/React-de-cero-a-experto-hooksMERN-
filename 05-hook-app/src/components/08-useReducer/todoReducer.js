import * as actions from './../../helpers/actions'
export const todoReducer = (state = [], action) => {
  switch (action?.type) {
    case actions.ADD_TODO:
      return [...state, action.payload]
    case actions.DEL_TODO:
      return state.filter(todo => todo.id !== action.payload)
    case actions.TOG_TODO:
      return state.map(todo => {
        if (todo.id === action.payload) {
          return { ...todo, done: !todo.done }
        } else {
          return todo
        }
      })
    case actions.TOGGLE_TODO:
      return state.map(todo =>
        todo.id === action.payload ? { ...todo, done: !todo.done } : todo
      )
    default:
      return state
  }
}
