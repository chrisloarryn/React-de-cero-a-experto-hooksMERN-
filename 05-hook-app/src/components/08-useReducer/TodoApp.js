import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer'

import './styles.css'
import { TodoList } from './components/TodoList'
import { TodoAdd } from './components/TodoAdd'

const init = () => JSON.parse(localStorage.getItem('todos')) || []

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init)

  console.log(todos)
  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])
  const handleDelete = todoId => {
    // create action
    const action = {
      type: 'DEL',
      payload: todoId,
    }
    // dispatch
    dispatch(action)
  }
  const handleToggle = todoId => dispatch({ type: 'TOGGLE', payload: todoId })

  const handleAddTodo = newTodo => dispatch({ type: 'ADD', payload: newTodo })

  return (
    <div>
      <h1>TodoApp ( {(todos && todos.length) || 0} )</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  )
}
