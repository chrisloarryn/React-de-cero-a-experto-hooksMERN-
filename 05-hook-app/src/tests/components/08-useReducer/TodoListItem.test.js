import React from 'react'
import { demoTodos } from './../fixtures/demoTodos'
import { shallow } from 'enzyme'
import { TodoListItem } from 'components/08-useReducer/TodoListItem'
import '@testing-library/jest-dom'

describe('Tests in <TodoListItem />', () => {
  const idx = 1
  const handleDelete = jest.fn()
  const handleToggle = jest.fn()

  let wrapper = shallow(
    <TodoListItem
      todo={demoTodos[idx]}
      index={idx}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
    />
  )

  beforeEach(() => {
    jest.clearAllMocks()
    wrapper = shallow(
      <TodoListItem
        todo={demoTodos[idx]}
        index={idx}
        handleDelete={handleDelete}
        handleToggle={handleToggle}
      />
    )
  })
  test('Should be rendered correctly', () => {
    // snapshot
    expect(wrapper).toMatchSnapshot()
  })

  test('Should call handleDelete function', () => {
    // jest.fn()
    // toHaveBeenCalled
    // toHaveBeenCalledWith
    wrapper.find('button').simulate('click')
    expect(handleDelete).toHaveBeenCalledWith(demoTodos[1].id)
  })

  test('Should call handleToggle function', () => {
    // jest.fn()
    // toHaveBeenCalled
    // toHaveBeenCalledWith
    wrapper.find('p').simulate('click')
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[1].id)
  })

  test('Should show text correctly', () => {
    // paragraph content
    const p = wrapper.find('p')
    expect(p.text().trim()).toBe(`${idx + 1} - ${demoTodos[idx].desc}`)
  })

  test('Should have complete class', () => {
    // paragraph content
    const todo = demoTodos[1]
    todo.done = true

    const wrapper = shallow(<TodoListItem todo={demoTodos[idx]} />)
    expect(wrapper.find('p').hasClass('complete')).toBe(true)
  })
})
