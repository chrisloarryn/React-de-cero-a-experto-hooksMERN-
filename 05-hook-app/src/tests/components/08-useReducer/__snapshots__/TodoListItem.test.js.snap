// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoListItem /> Should be rendered correctly 1`] = `
<li
  className="list-group-item"
  key="2"
>
  <p
    className="false"
    onClick={[Function]}
  >
    2 - Learn Mongo
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
